test/metabase/sync_database_test.clj

```diff
@@ -405,3 +414,110 @@
+   (:fields_hash (db/select-one Table :id table-id))])
+
+;; This tests a table that adds, drops and readds a column to ensure that we know somethings changed, sync and update
+;; the hash
+(expect
+  [
+   ;; Original number of active/inactive fields
+   [1 0]
+   ;; Add a column, should still be no inactive
+   [2 0]
+   ;; Adding a column should make the hashes not equal
+   false
+   ;; Dropped the added column
+   [1 1]
+   ;; Dropping the column should make the hash not equal to the previous
+   false
+   ;; The hash should be the same as the first one though
+   true
+   ;; Reactive the dropped column
+   [2 0]
+   ;; Should trigger a change to the hash
+   false
+   ;; Reactivating the column should cause the hash to be the same as before it was dropped
+   true
+   ]
+  (let [details {:db (str "mem:" (tu/random-name) ";DB_CLOSE_DELAY=10")}]
+    (binding [mdb/*allow-potentailly-unsafe-connections* true]
+      (tt/with-temp Database [db {:engine :h2, :details details}]
+        (jdbc/with-db-connection [conn (sql/connection-details->spec (driver/engine->driver :h2) details)]
+          (let [get-table #(db/select-one Table :db_id (u/get-id db))]
+            ;; create the `blueberries_consumed` table and insert 50 values
+            (exec! conn ["CREATE TABLE blueberries_consumed (num SMALLINT NOT NULL)"
+                         (insert-range-sql (range 50))])
+            (sync-database! db)
+            ;; We should now have a hash value for num as a SMALLINT
+            (let [table-id                  (u/get-id (get-table))
+                  [fc-1 inactive-fc1 hash1] (table-md-with-hash table-id)]
+              (exec! conn ["ALTER TABLE blueberries_consumed ADD COLUMN weight FLOAT"])
+              (sync-database! db)
+              ;; Now that has will include num and weight
+              (let [[fc-2 inactive-fc2 hash2] (table-md-with-hash table-id)]
+                (exec! conn ["ALTER TABLE blueberries_consumed DROP COLUMN weight"])
+                (sync-database! db)
+                ;; We no longer have weight in the hash, so we should have the same hash as before with only num
+                (let [[fc-3 inactive-fc3 hash3] (table-md-with-hash table-id)]
+                  (exec! conn ["ALTER TABLE blueberries_consumed ADD COLUMN weight FLOAT"])
+                  (sync-database! db)
+                  ;; Adding back in weight should change the hash again, but it should be the same as the one before
+                  ;; that included num and weight
+                  (let [[fc-4 inactive-fc4 hash4] (table-md-with-hash table-id)]
+                    [[fc-1 inactive-fc1]
+                     [fc-2 inactive-fc2]
+                     (= hash1 hash2)
+                     [fc-3 inactive-fc3]
+                     (= hash2 hash3)
+                     (= hash1 hash3)
+                     [fc-4 inactive-fc4]
+                     (= hash3 hash4)
+                     (= hash2 hash4)]))))))))))
```

Comment:
this test makes my head hurt.
Is there no way to break out some of the different parts into different tests?

Also FWIW instead of putting things in a vector I think doing something like 

```clj
(array-map
  :1 ...
  :2 ...
  :3 ...)
```

and having `expected` be

```clj
{:1 ...
 :2 ...
 :3 ...}
```

would make this a little more readable

Reply:
although not sticking to many different tests in a single test in the first place is even better

Reply:
I agree about there being too much going on in that test, I'll try and split some of those scenarios to simplify it.